scanner/client/client.go

```diff
@@ -0,0 +1,190 @@
+package client
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"time"
+
+	"github.com/cenkalti/backoff/v3"
+	"github.com/google/go-containerregistry/pkg/authn"
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/quay/zlog"
+	v4 "github.com/stackrox/rox/generated/internalapi/scanner/v4"
+	"github.com/stackrox/rox/pkg/clientconn"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// ScannerClient is the interface that contains the StackRox Scanner
+// application-oriented methods. It's offered to simplify application code to
+// call StackRox Scanner:
+//
+//   - GetOrCreateImageIndex first attempts to get an existing index report for the
+//     image reference, and if not found or invalid, it then attempts to index the
+//     image and return the generated index report if successful, or error.
+//
+//   - IndexAndScanImage scans an image for vulnerabilities. If the index report
+//     for that image does not exist, it is created. It returns the vulnerability
+//     report.
+//
+//   - Close cleans up any resources used by the implementation.
+type ScannerClient interface {
```

Comment:
Do you mean in the implementation or on the interface?

I am unfamiliar with how IDEs parse the docstrings to achieve your suggestion. I want to support that, but can you tell me the generic go-lang-supported way to do that?

The one I am proposing in the patch I copied from standard libraries, by the way.

Reply:
It depends on the `type` of the variable holding the `client`, in this case since `NewGRPCScannerClient` returns `ScannerClient` (as opposed to `grpcClient`) I would put the descriptions on the `ScannerClient` interface methods.

The comments in the `main` method of snippet below are what `vscode` shows when hovering - not sure if other IDEs do the same.

```go
package main

// INTERFACE-TYPE
type BlahInterface interface {
	// INTERFACE-FUNC
	BlahMethod()
}

// IMPL-TYPE
type blahImpl struct{}

// IMPL-FUNC
func (b *blahImpl) BlahMethod() {}

func NewBlahInter() BlahInterface {
	return &blahImpl{}
}

func NewBlahImpl() *blahImpl {
	return &blahImpl{}
}

func main() {
	var inter BlahInterface  // INTERFACE-TYPE and INTERFACE-FUNC
	inter.BlahMethod()       // INTERFACE-FUNC

	inter = NewBlahInter()   
	inter.BlahMethod()       // INTERFACE-FUNC

	inter = NewBlahImpl()    
	inter.BlahMethod()       // INTERFACE-FUNC

	inter = &blahImpl{}      // IMPL-TYPE
	inter.BlahMethod()       // INTERFACE-FUNC

	var impl *blahImpl       // IMPL-TYPE
	impl.BlahMethod()        // IMPL-FUNC

	impl = NewBlahImpl()       
	impl.BlahMethod()        // IMPL-FUNC
}

```

Reply:
I was browsing the standard library and found examples of both, like [these two interfaces](https://github.com/golang/go/blob/1f4e8afafee02b69231ac31bfff7c7211bb41336/src/database/sql/driver/driver.go#L215-L254) - one has the `interface` level comment, the other also has the `func` level comments. Based on that example perhaps if the `func` name is descriptive enough then omission is warranted. Do you have a link to standard lib example you are referring to? Wondering if this is something we should add to our style guide / checks.

Reply:
fd2495fcd3..90c02f0ffe